homeassistant/components/tesla_wall_connector/__init__.py

```diff
@@ -0,0 +1,190 @@
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    CONF_SCAN_INTERVAL_CHARGING,
+    DEFAULT_SCAN_INTERVAL,
+    DEFAULT_SCAN_INTERVAL_CHARGING,
+    DOMAIN,
+    WALLCONNECTOR_CLIENT,
+    WALLCONNECTOR_DATA_LIFETIME,
+    WALLCONNECTOR_DATA_UPDATE_COORDINATOR,
+    WALLCONNECTOR_DATA_VITALS,
+    WALLCONNECTOR_DEVICE_NAME,
+    WALLCONNECTOR_FIRMWARE_VERSION,
+    WALLCONNECTOR_HOST,
+    WALLCONNECTOR_PART_NUMBER,
+    WALLCONNECTOR_SERIAL_NUMBER,
+)
+
+PLATFORMS: list[str] = ["sensor", "binary_sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Tesla Wall Connector from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    hostname = entry.data[CONF_HOST]
+    poll_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    poll_interval_charging = entry.options.get(
+        CONF_SCAN_INTERVAL_CHARGING, DEFAULT_SCAN_INTERVAL_CHARGING
+    )
+
+    wall_connector = WallConnector(host=hostname, session=async_get_clientsession(hass))
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        WALLCONNECTOR_CLIENT: wall_connector,
+        WALLCONNECTOR_HOST: hostname,
+    }
+
+    try:
+        version_data = await wall_connector.async_get_version()
+    except WallConnectorError as ex:
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id].update(
+        {
+            WALLCONNECTOR_PART_NUMBER: version_data.part_number,
+            WALLCONNECTOR_FIRMWARE_VERSION: version_data.firmware_version,
+            WALLCONNECTOR_SERIAL_NUMBER: version_data.serial_number,
+        }
+    )
+
+    async def async_update_data():
+        """Fetch new data from the Wall Connector."""
+        wall_connector = hass.data[DOMAIN][entry.entry_id][WALLCONNECTOR_CLIENT]
+        coordinator = hass.data[DOMAIN][entry.entry_id][
+            WALLCONNECTOR_DATA_UPDATE_COORDINATOR
+        ]
+
+        try:
+            previous_contactor_closed = (
+                coordinator.data[WALLCONNECTOR_DATA_VITALS].contactor_closed
+                if coordinator.data is not None
+                and coordinator.data[WALLCONNECTOR_DATA_VITALS] is not None
+                else False
+            )
+            vitals = await wall_connector.async_get_vitals()
+            lifetime = await wall_connector.async_get_lifetime()
+
+            if previous_contactor_closed != vitals.contactor_closed:
+                coordinator.update_interval = (
+                    timedelta(seconds=poll_interval)
+                    if not vitals.contactor_closed
+                    else timedelta(seconds=poll_interval_charging)
+                )
+                _LOGGER.debug(
+                    "Contactor closed: %s. Update interval: %s",
+                    str(vitals.contactor_closed),
+                    str(coordinator.update_interval),
+                )
```

Comment:
Move code out of the try block that isn't expected to throw

Reply:
Fixed in 99683d4